peloton/update_project.py

```diff
@@ -0,0 +1,1371 @@
+
+    def _run_sub_mutation(self, page_slice: slice) -> list:
+        """
+        Build, run and return results from a mutation for one page of fields.
+        """
+        mutation = sgqlc.operation.Operation(github_schema_root.mutation_type)
+        for ix in self._indexes[page_slice]:
+            self._add_selector(mutation=mutation, index=ix)
+        result = run_operation(mutation)
+        # Fields are accessed from mutation results via their keys.
+        payloads = [result[a] for a in self._aliases[page_slice]]
+        return payloads
+
+    def _get_data_from_sub_mutations(self, payloads: list[list]) -> list | None:
+        """
+        Extract data from the nested results list that a mutation returns.
+
+        Provided for mutations that need to return something e.g. IDs of
+        issues newly added to the project. Otherwise: no need to subclass.
+        """
+        return None
+
+
+class AddIssuesMutation(PaginatedMutation):
+    """
+    Mutation to add Issues to the Peloton GitHub project.
+
+    Added via the IDs of the issues itself, and records (self.results) the
+    IDs of the corresponding items in the project.
+    """
+    def _make_inputs(
+        self,
+        data_frame: pd.DataFrame,
+        project_cols: ProjectItemsQuery.ColNames,
+        issue_cols: IssuesQuery.ColNames,
+    ) -> list[dict]:
+        return [
+            # Expecting a DataFrame where the required column is the index.
+            (self._inputs_default | dict(content_id=row.name))
+            for _, row in data_frame.iterrows()
+        ]
+
+    def _add_selector(
+        self,
+        mutation: sgqlc.operation.Operation,
+        index: int
+    ) -> None:
+        add = mutation.add_project_v2_item_by_id(
+            input=self._inputs[index], __alias__=self._aliases[index]
+        )
+        add.item().id()
+
+    def _get_data_from_sub_mutations(
+        self,
+        payloads: list[list[github_schema.AddProjectV2ItemByIdPayload]]
+    ) -> list[str]:
+        result = []
+        for sub_list in payloads:
+            ids_in_project = [p.item.id for p in sub_list]
+            result.extend(ids_in_project)
+        return result
+
+
+class AddDraftsMutation(PaginatedMutation):
+    """
+    Mutation to add Discussions to the Peloton GitHub project.
+
+    Added by adding project 'drafts' - via a given title and body text - and
+    records (self.results) the IDs of the corresponding items in the project.
+    """
+    def _make_inputs(
+        self,
+        data_frame: pd.DataFrame,
+        project_cols: ProjectItemsQuery.ColNames,
+        issue_cols: IssuesQuery.ColNames,
+    ) -> list[dict]:
+        inputs = []
+        for _, row in data_frame.iterrows():
+            title = row[issue_cols.title]
+            # TODO: how to avoid bad unicode escape sequence with emojis?
+            #  sgqlc uses bytes(op).decode('utf-8') on its queries which
+            #  might be the
+            #  source of the problem.
```

Comment:
@trexfeathers Turn this into an issue?

Might be a way to tease other devs to contribute to the code-base ...

Reply:
Sounds good. Bug me about this one - I can't make a meaningful issue until this code is merged.